src/bit_bang_i2c.cpp

```diff
@@ -1,95 +1,112 @@
 #include <chrono>
 
-#include <libhal-soft/i2c_bit_bang.hpp>
+#include <libhal-soft/bit_bang_i2c.hpp>
 #include <libhal-util/bit.hpp>
 #include <libhal-util/i2c.hpp>
 #include <libhal-util/steady_clock.hpp>
 #include <libhal/error.hpp>
 #include <libhal/units.hpp>
 
 namespace hal {
-// public
-i2c_bit_bang::i2c_bit_bang(i2c_pins p_pins,
+
+/**
+ * @brief This function is a high speed version of the hal::delay function
+ * which operates on ticks
+ *
+ * @param ticks The amount of ticks this function will delay for
+ */
+
+[[gnu::always_inline]] inline void high_speed_delay(
+  steady_clock* p_steady_clock,
+  uint64_t ticks)
+{
+  const auto start_time_high = p_steady_clock->uptime();
+  uint64_t uptime = 0;
+
+  const auto ticks_until_timeout_high = ticks + start_time_high;
+
+  while (uptime < ticks_until_timeout_high) {
+    uptime = p_steady_clock->uptime();
+    continue;
+  }
+}
+
+// Public
+bit_bang_i2c::bit_bang_i2c(pins p_pins,
                            steady_clock& p_clock,
-                           bus_info p_bus)
+                           const float p_duty_cycle)
   : m_scl(p_pins.scl)
   , m_sda(p_pins.sda)
   , m_clock(&p_clock)
-  , m_bus(p_bus)
+  , m_duty_cycle(p_duty_cycle)
 {
 }
 
-void i2c_bit_bang::driver_configure(const settings& p_settings)
+// Private
+
+void bit_bang_i2c::driver_configure(const settings& p_settings)
 {
   using namespace std::chrono_literals;
 
   if (p_settings.clock_rate > m_clock->frequency()) {
     throw hal::operation_not_supported(this);
   }
 
-  // calculate period in nanosecond
   using period = std::chrono::nanoseconds::period;
-  auto period_ns = hal::wavelength<period>(p_settings.clock_rate);
-  auto scl_high_time = period_ns * m_bus.duty_cycle;
-  auto scl_low_time = period_ns - scl_high_time;
-
-  // calculate ticks for high and low
-  const auto frequency = m_clock->frequency();
-  const auto tick_period = hal::wavelength<period>(frequency);
-
-  m_scl_high_ticks = static_cast<uint64_t>(scl_high_time / tick_period);
 
-  m_scl_low_ticks = static_cast<uint64_t>(scl_low_time / tick_period);
-
-  // calculate the delay due to the uptime function call
+  // Calculate the delay due to the uptime function call
   const auto callibration_start_tick = m_clock->uptime();
   const auto callibration_end_tick = m_clock->uptime();
-  m_uptime_ticks = callibration_end_tick - callibration_start_tick;
+  const auto uptime_ticks = callibration_end_tick - callibration_start_tick;
 
-  // calculating output_pin going to true delay time
-  const auto before_output_high = m_clock->uptime();
+  // Calculating output_pin going to true delay time
+  const auto before_output = m_clock->uptime();
   m_scl->level(true);
-  const auto after_output_high = m_clock->uptime();
-  const auto output_true_ticks = after_output_high - before_output_high;
+  const auto after_output = m_clock->uptime();
+  const auto calibration_ticks = after_output - before_output - uptime_ticks;
```

Comment:
I'm concerned that this may become negative, which would add to the m_scl_high_ticks, can you check the math to ensure its valid.

Reply:
I will double check it and explain my reasonings behind this calibration.

Reply:
So my original code before we decided to change over to this was adding the uptime_ticks from the uptime call every time to check whether or not the desired amount of ticks was achieved. But we wanted to save some computation time and decided to move it out of the function and the only way to do that was to move it to this variable which would equate to being on the other side of the comparison for the high_speed_delay function which would also mean that I need to subtract it from this variable.

Reply:
This is interesting. I would say this shouldn't happen but there is a chance that it could if something happens external to the system which significantly slows down the first two calls. But considering we do 2 uptime calls and store that in uptime_ticks then we do another 2 uptime calls PLUS a level call to a output pin, this value should always be greater then uptime call. But looking lower, to line 77-81 there is another spot for a possible negative value if the user for some reason passes in a 0 frequency.

Reply:
Upon additional thinking on this subject. I don't think that a negative value on these calibrations would be inherently terrible, this would just indicate that the driver is attempting to handle frequencies faster then what bit bang can deal with. I think this is interesting because if you are attempting to achieve to high of frequencies then you shouldn't be utilizing this interface in the first place. You should have said implementation of i2c. However, I can put some checks in here and throw some errors if the system sees that the operations here will force it into a negative state, which would mean throwing a hal::operation_not_supported.

Reply:
@kammce

Reply:
Hmmm, I realize there may also be one more flaw which is an interrupt occurring in between these calls. Such an event would cause the time between calls to be unspecified in length. 

Having the speed be too fast is an issue if the devices on the bus cannot support it.

Reply:
Calibration may actually be a bad idea here because of this issue. We cannot guarantee anything about the calibration time, unless we used a critical section (disabled interrupts). So we may need to strip this out because of that. Sorry about that. I hadn't considered this until now.

Reply:
Agreed. Calibration was found to only increase accuracy of the frequency a negligible amount and introduced more considerations for the end user. Therefore we agreed to remove it from the driver.

Reply:
This is the old code for the calibration
`
  using namespace std::chrono_literals;

  if (p_settings.clock_rate > m_clock->frequency()) {
    throw hal::operation_not_supported(this);
  }

  using period = std::chrono::nanoseconds::period;

  // Calculate the delay due to the uptime function call
  const auto callibration_start_tick = m_clock->uptime();
  const auto callibration_end_tick = m_clock->uptime();
  const auto uptime_ticks = callibration_end_tick - callibration_start_tick;

  // Calculating output_pin going to true delay time
  const auto before_output = m_clock->uptime();
  m_scl->level(true);
  const auto after_output = m_clock->uptime();
  const auto calibration_ticks = after_output - before_output - uptime_ticks;

  // Calculate period in nanosecond
  auto period_ns = hal::wavelength<period>(p_settings.clock_rate);
  auto scl_high_time = period_ns * m_duty_cycle;
  auto scl_low_time = period_ns - scl_high_time;

  // Calculate ticks for high and low
  const auto frequency = m_clock->frequency();
  const auto tick_period = hal::wavelength<period>(frequency);

  m_scl_high_ticks =
    static_cast<uint64_t>(scl_high_time / tick_period) - calibration_ticks;

  m_scl_low_ticks =
    static_cast<uint64_t>(scl_low_time / tick_period) - calibration_ticks;`